src/main/java/uk/gov/pay/directdebit/partnerapp/services/GoCardlessAppConnectAccountService.java

```diff
@@ -0,0 +1,85 @@
+    private final GoCardlessAppConnectClient connectClient;
+
+    @Inject
+    public GoCardlessAppConnectAccountService(GoCardlessAppConnectAccountTokenDao goCardlessAppConnectAccountTokenDao,
+                                              GatewayAccountDao gatewayAccountDao,
+                                              GoCardlessAppConnectClient connectClient) {
+        this.goCardlessAppConnectAccountTokenDao = goCardlessAppConnectAccountTokenDao;
+        this.gatewayAccountDao = gatewayAccountDao;
+        this.connectClient = connectClient;
+    }
+
+    public Response createToken(String gatewayAccountExternalId,
+                                String redirectUri) {
+
+        return gatewayAccountDao.findByExternalId(gatewayAccountExternalId)
+                .map(account -> goCardlessAppConnectAccountTokenDao.findByGatewayAccountId(account.getId())
+                        .map(token -> {
+                            goCardlessAppConnectAccountTokenDao.disableToken(token.getToken(), account.getId());
+                            return mapEntity(insertToken(account, redirectUri));
+                        })
+                        .orElseGet(() -> mapEntity(insertToken(account, redirectUri))))
+                .orElseThrow(() -> new BadRequestException("There is no gateway account with external id " + gatewayAccountExternalId));
+    }
+
+    public Response exchangeCodeForToken(String accessCode, String partnerToken) {
+        return goCardlessAppConnectAccountTokenDao.findActiveTokenByToken(partnerToken)
+                .map(token -> gatewayAccountDao.findById(token.getGatewayAccountId())
+                        .map(gatewayAccount -> connectClient.postAccessCode(accessCode, gatewayAccount, token.getRedirectUri())
+                                .map(clientResponse -> processGoCardlessResponse(clientResponse, gatewayAccount))
+                                .orElseThrow(() -> new InternalServerErrorException("There is no response from GoCardless Connect client")))
+                        .orElseThrow(() -> new BadRequestException("There is no gateway account with id " + token.getGatewayAccountId().toString())))
+                .orElseThrow(() -> new BadRequestException("There is no token with value " + partnerToken));
+    }
+
+    private Response processGoCardlessResponse(GoCardlessAppConnectAccessTokenResponse response, GatewayAccount gatewayAccount) {
+        if (GoCardlessAppConnectClient.isValidResponse(response)) {
+            gatewayAccountDao.updateAccessTokenAndOrganisation(gatewayAccount.getExternalId(), response.getAccessToken(), response.getOrganisationId());
+            return Response.ok().build();
+        } else if (StringUtils.isNotBlank(response.getError())) {
+            return Response.status(400).entity(GoCardlessAppConnectCodeExchangeErrorResponse.from(response)).build();
+        } else {
+            throw new BadRequestException("Received and invalid response from GoCardless Connect");
```

Comment:
I don’t think this is a 400 Bad Request because there the problem is with GoCardless rather than the request to Direct Debit connector (assuming our validation etc. is good enough). Probably a 500 Internal Server Error.

Reply:
👍